File: Java-CodeServer/src/main/resources/js/prompto/literal/HexaLiteral.js
var Literal = require("./Literal").Literal;
var Integer = require("../value/Integer").Integer;
var IntegerType = require("../type/IntegerType").IntegerType;

/* jshint bitwise:false*/
function parseHexa(text) {
	var value = 0;
	for(var i=2;i<text.length; i++) {
		value <<= 4;
		var c = text[i];
		if(c>='0' && c<='9') {
			value += (c.charCodeAt(0) - '0'.charCodeAt(0));
		} else if(c>='a' && c<='f') {
			value += (c.charCodeAt(0) - 'a'.charCodeAt(0));
		} else if(c>='A' && c<='F') {
			value += 10 + (c.charCodeAt(0) - 'A'.charCodeAt(0));
		} else {
			throw (text + " is not a valid hexadecimal number");
		}
	}
	return new Integer(value);
}

function HexaLiteral(text) {
	Literal.call(this, text, parseHexa(text));
	return this;
}

HexaLiteral.prototype = Object.create(Literal.prototype);
HexaLiteral.prototype.constructor = HexaLiteral;

HexaLiteral.parseHexa = parseHexa;
HexaLiteral.prototype.check = function(context) {
	return IntegerType.instance;
};

exports.HexaLiteral = HexaLiteral;

File: Core/src/main/java/prompto/literal/HexaLiteral.java
package prompto.literal;

import prompto.runtime.Context;
import prompto.type.IType;
import prompto.type.IntegerType;
import prompto.value.IntegerValue;

public class HexaLiteral extends Literal<IntegerValue> {

	
	public HexaLiteral(String text) {
		super(text,parseHexa(text));
	}


	@Override
	public IType check(Context context) {
		return IntegerType.instance();
	}
	
	
	static public IntegerValue parseHexa(String text) {
		long value = 0;
		for(char c : text.substring(2).toCharArray()) {
			value <<= 4;
			if(c>='0' && c<='9')
				value += (c - '0');
			else if(c>='a' && c<='f')
				value += (c - 'a');
			else if(c>='A' && c<='F')
				value += 10 + (c - 'A');
			else
				throw new NumberFormatException(text + " is not a valid hexadecimal number");
		}
		return new IntegerValue(value);
	}
	
}
